<!DOCTYPE html>
<html prefix="og: http://ogp.me/ns# article: http://ogp.me/ns/article# " lang="en">
<head>
<meta charset="utf-8">
<base href="http://kmdouglass.github.io/index-1.html">
<meta name="description" content="Biophysics, optics, and programming.">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Kyle M. Douglass (old posts, page 1) | Kyle M. Douglass</title>
<link href="assets/css/all-nocdn.css" rel="stylesheet" type="text/css">
<meta content="#5670d4" name="theme-color">
<link rel="alternate" type="application/rss+xml" title="RSS" href="rss.xml">
<link rel="canonical" href="http://kmdouglass.github.io/index-1.html">
<link rel="prev" href="index.html" type="text/html">
<!--[if lt IE 9]><script src="assets/js/html5.js"></script><![endif]--><script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
</head>
<body>
<a href="#content" class="sr-only sr-only-focusable">Skip to main content</a>

<!-- Menubar -->

<nav class="navbar navbar-default navbar-static-top"><div class="container">
<!-- This keeps the margins nice -->
        <div class="navbar-header">
            <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-navbar" aria-controls="bs-navbar" aria-expanded="false">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand" href="http://kmdouglass.github.io/">

                <span id="blog-title">Kyle M. Douglass</span>
            </a>
        </div>
<!-- /.navbar-header -->
        <div class="collapse navbar-collapse" id="bs-navbar" aria-expanded="false">
            <ul class="nav navbar-nav">
<li>
<a href="archive.html">Archive</a>
                </li>
<li>
<a href="categories/index.html">Tags</a>
                </li>
<li>
<a href="stories/notes-and-projects.html">Notes</a>
                </li>
<li>
<a href="stories/about-me.html">About</a>
                </li>
<li>
<a href="stories/contact.html">Contact</a>
                </li>
<li>
<a href="rss.xml">RSS feed</a>

                
            </li>
</ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!-- /.navbar-collapse -->
    </div>
<!-- /.container -->
</nav><!-- End of Menubar --><div class="container" id="content" role="main">
    <div class="body-content">
        <!--Body content-->
        <div class="row">
            
            

<div class="postindex">
    <article class="h-entry post-text"><header><h1 class="p-name entry-title"><a href="posts/measurements.html" class="u-url">Measurements as processes</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">
                Kyle M. Douglass
            </span></p>
            <p class="dateline"><a href="posts/measurements.html" rel="bookmark"><time class="published dt-published" datetime="2014-12-19T00:00:00+01:00" title="2014-12-19 00:00">2014-12-19 00:00</time></a></p>
                <p class="commentline">
        
    <a href="posts/measurements.html#disqus_thread" data-disqus-identifier="cache/posts/measurements.html">Comments</a>


        </p>
</div>
    </header><div class="p-summary entry-summary">
    <p>
I work in microscopy, which is one form of optical sensing. In the
sensing field, we are often concerned with making measurements on
some structure so as to learn what it is. Often, I think the word
measurement refers to the dataset that's produced.
</p>

<p>
I think it can be more effective to think of a measurement as a
process that transforms the structure into the dataset. Why is this
so? Well, to understand what the original structure was, we have to
look at our data and make an inference. If we understand the steps
in the process that took the original structure and turned it into
data, we can apply the inverse of those steps in reverse order to
get the original structure.
</p>

<p>
Of course, our dataset may only capture some limited aspects of the
original structure, so we may only be able to make probabilistic
statements about what the original structure was.
</p>

<p>
Take, for example, super-resolution microscopy experiments (SR). In
SR, some feature of a cell is labeled with a discrete number of
fluorescent molecules, then these molecules are localized to a high
precision. The centroids of all the molecules are then convolved
with a Gaussian function (or something similar) with a width equal
to the localization precision to produce a rendered, super-resolved
image of the structure. The measurement process can be thought of
like this:
</p>

<ol class="org-ol">
<li>Attach fluorescent molecules to every macromolecule (or randomly
to a subset of macromolecules) in the structure of interest.
</li>
<li>For every molecule that emits photons during the time of
acqusition by one camera frame, record its true coordinate
positions and the number of detected photons. This can create
multiple localizations that correspond to the same molecule in
multiple camera frames.
</li>
<li>Remove molecules from the lis that emitted less than some
threshold number of photons. These correspond to molecules with a
signal-to-noise ratio that is too low to be detected.
</li>
<li>Randomly bump the molecule positions according to a Gaussian
distribution with width equal to the localization precision in
each direction.
</li>
</ol>
<p>
This process results in a list of molecule positions that originally
were located on the structure of interest, but were eventually
displaced randomly and filtered out due to various sources of
noise. To understand what the original structure was, we have to
"undo" each of these steps to the best of our abilities.
</p>

<p>
I think it's interesting to note that everytime a random change to
the original molecule positions occurred, we lose some information
about the structure.
</p>
    </div>
    </article><article class="h-entry post-text"><header><h1 class="p-name entry-title"><a href="posts/confocal-by-scattering.html" class="u-url">Parallelized confocal microscopy with multiply scattered light</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">
                Kyle M. Douglass
            </span></p>
            <p class="dateline"><a href="posts/confocal-by-scattering.html" rel="bookmark"><time class="published dt-published" datetime="2014-10-09T00:00:00+02:00" title="2014-10-09 00:00">2014-10-09 00:00</time></a></p>
                <p class="commentline">
        
    <a href="posts/confocal-by-scattering.html#disqus_thread" data-disqus-identifier="cache/posts/confocal-by-scattering.html">Comments</a>


        </p>
</div>
    </header><div class="p-summary entry-summary">
    <p>
My PhD work dealt with the topic of sensing and imaging using light
that had been transmitted through a random medium. This topic is
often applied to practical problems such as imaging through a
turbulent atmosphere or detecting objects, like a tumor, buried
beneath an opaque substance, like skin and muscle tissue. Though I
don't necessarily work in this field anymore, I still follow its
developments occasionally.
</p>

<p>
A few articles appeared within the last week in journals like <a href="http://www.nature.com/nphoton/journal/v8/n10/full/nphoton.2014.189.html">Nature</a>
Photonics]] about the problem of imaging through walls. This problem
has been studied since the late 1980's in a number of papers,
including the notable <a href="http://journals.aps.org/prl/abstract/10.1103/PhysRevLett.61.834">Feng, Kane, Lee, and Stone</a> paper and <a href="http://www.sciencedirect.com/science/article/pii/037596019090615U#">Freund's
discussion</a> of using cross-correlating two speckle patterns, one of
which is a reference wave, to do a sort of holographic imaging. The
recent work continues from these original ideas and applies them to
microscopy.
</p>

<p>
One article appeared on the arXiv and is from the <a href="http://arxiv.org/abs/1410.2079">Mosk and Lagendijk</a>
camp. In this article, they exploit the optical memory effect,
whereby a speckle pattern generated by the multiple scattering of a
plane wave by a random slab is simply translated as the angle of
incidence of the plane wave is varied. If a thin fluorescent target
is placed directly on the opposite face of the scattering slab, it
will be excited by the speckle pattern and emit a fluorescence
signal that can be captured by an objective. Changing the angle of
incidence of the plane wave then allows for multiple points to scan
the sample in parallel. Ultimately, a number of images are taken
with the sample illuminated by several transversally shifted speckle
patterns and the resulting 4D data cube (corresponding to the
sample's x-y dimensions and the two tilt angles of the incident
plane wave) is used to render an image with improved resolution.
</p>

<p>
As stated in the article's title, the resolution improvement is
relative to that of a widefield microscope. They obtain an effective
point spread function of about 140 nm and a field of view of 10
microns by 10 microns.
</p>

<p>
So how does the technique work? My first thought was that the memory
effect is simply another way of saying that the speckle pattern acts
as a number of confocal-like point sources, which essentially means
this is something of a parallelized confocal microscope. However,
I'm not sure this is correct for two reasons: 1) there is no
detection pin hole, and 2) the angle of incidence of the plane wave
is scanned over a small angular range so that the speckle pattern is
simply translated. If the angle of incidence is so great that the
linear change in phase of the plane wave is greater than roughly the
size of the scattering slab, the speckle pattern is no longer simply
translated but changes completely.
</p>

<p>
In reality, the ultimate resolution of this technique is the average
speckle grain size, which can't be less than about half the
wavelength. This suggests that the angular spectrum of the speckle
pattern is what determines the resolution improvement.
</p>

<p>
The speckle pattern in a region bounded by the maximum extent of the
memory effect has a fixed angular spectrum and translating the
speckle pattern only changes the phase of the spectrum. So, scanning
a target with a speckle pattern produces beat patterns containing
high spatial frequency information that can propgate on the low
spatial frequency waves that reach the objective. Translating the
speckle pattern then performs a sort of phase-shifting
interferometry that allows for both intensity and phase retrieval of
the object.
</p>

<p>
Importantly, if the speckle pattern is scanned outside the range of
the memory effect, the speckle's angular spectrum within the region
changes completely so that the original reference wave is lost. The
fact that the object is fluorescent and not simply scattering the
light shouldn't matter if the fluorescence intensity is linear with
the excitation light intensity. However, if the fluorescence has
saturated at the average intensity of the speckle pattern, then I'm
not exactly sure that this technique will work (though maybe some
sort of nonlinear structured illumination microscopy could be
achieved).
</p>

<p>
Overall it's a neat demonstration and worth the exercise to
understand it, though I'm doubtful that at this point it would be
useful for applications in the life sciences. This is because the
resolution isn't that much better than a spinning disk confocal
microscope, which has a larger field of view and would arguably be
easier to use by biologists.
</p>
    </div>
    </article><article class="h-entry post-text"><header><h1 class="p-name entry-title"><a href="posts/half-wave%20plate.html" class="u-url">An accidental half-wave plate</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">
                Kyle M. Douglass
            </span></p>
            <p class="dateline"><a href="posts/half-wave%20plate.html" rel="bookmark"><time class="published dt-published" datetime="2014-08-06T00:00:00+02:00" title="2014-08-06 00:00">2014-08-06 00:00</time></a></p>
                <p class="commentline">
        
    <a href="posts/half-wave%20plate.html#disqus_thread" data-disqus-identifier="cache/posts/half-wave plate.html">Comments</a>


        </p>
</div>
    </header><div class="p-summary entry-summary">
    <p>
Recently in the lab we made a seemingly minor change to one of our
microscopes that led to a minor problem that lasted for about a
week. Briefly, we introduced a mirror following a periscope that
raises the plane that the laser beams travel in parallel to the
table. The periscope is necessary to bringthe beams out of the plane
containing the beam-combining optics and into the TIRF module port
on our Zeiss inverted microscope for epi-illumination.
</p>

<p>
We had introduced the mirror to give us one more degree of freedom
for steering the beam, but to do so we had to move the periscope and
turn one of the mirrors by 90 degrees. Unfortunately, after doing
this we found that the optical power leaving the microscope
objective had dropped by a factor of 10, which was insufficient to
do STORM imaging.
</p>

<p>
We eventually determined that the Zeiss TIRF module contained a
polarizing beam splitter that combined the laser beams with white
light from another port at 90 degrees to the beam path. In the
newsetup, we placed a broadband halfwave plate before the TIRF
module, rotated it while watching a power meter, and were finally
able to get even more power leaving the objective than with the old
setup.
</p>

<p>
So what had happened to cause the decrease in power by adding
another mirror, and why did introducing the halfwave plate fix the
problem? As it turns out, you can rotate the polarization by 90
degrees using only a pair of mirrors to redirect the beam into a
plane parallel to its original plane of travel but into a direction
perpendicular to its original one. This was what we were doing
before we introduced the mirror. After the new mirror was put in
place, we rotated a mirror of the periscope, which effectively
rotated the laser beam's polarization back into its original
orientation. And, since there was a polarizing beam splitter inside
the TIRF module, the new beam polarization was transmitted through
the module instead of being reflected towards the objective.
</p>

<p>
The picture below describes how this can happen. Note that the final
beam trajectory is in the direction pointing <i>into</i> the screen.
</p>


<div class="figure">
<p><img src="images/polarization_rotator.png" alt="polarization_rotator.png"></p>
</div>
    </div>
    </article><article class="h-entry post-text"><header><h1 class="p-name entry-title"><a href="posts/omics.html" class="u-url">The philosophy of -omics studies</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">
                Kyle M. Douglass
            </span></p>
            <p class="dateline"><a href="posts/omics.html" rel="bookmark"><time class="published dt-published" datetime="2014-04-01T00:00:00+02:00" title="2014-04-01 00:00">2014-04-01 00:00</time></a></p>
                <p class="commentline">
        
    <a href="posts/omics.html#disqus_thread" data-disqus-identifier="cache/posts/omics.html">Comments</a>


        </p>
</div>
    </header><div class="p-summary entry-summary">
    <p>
As a physcist, I'm a relative newcomer to the biological
sciences. As a result, many approaches to doing science within these
fields are new and really interesting to me.
</p>

<p>
One such idea is the <i>-omics</i> approach to science. Sonja Prohaska
and Peter Stadler provide an insightful and sometimes amusing
intrepetation of <i>-omics</i> studies in an article in the book
<i>Bioinformatics for Omics Data.</i> (The PubMed link to the article is
<a href="http://www.ncbi.nlm.nih.gov/pubmed/21370084">here</a>, but you can find a pdf by Googling "The use and abuse of
-omes.") According to them, <i>-omics</i> refers to the research field of
digesting the pile of data coming from measurements of a particular
<i>-ome</i>, such as the "genome" or "transcriptome." The goal is to
relate the collection of parts within the <i>-ome</i> to biological
function, or at least to determine function by comparing <i>-omes</i> of
two different organisms.
</p>

<p>
The authors explain that <i>-omics</i> approaches have three components,
which I quote from their article:
</p>

<ol class="org-ol">
<li>A suite of (typically high-throughput) technologies address a
well-defined collection of biological objects, the pertinent
<i>-ome</i>.
</li>
<li>Both technologically and conceptually there is a cataloging
effort to enumerate and characterize the individual
objects–hopefully coupled with an initiative to make this
catalog available as a database.
</li>
<li>Beyond the enumeration of objects, one strives to cover the
quantitative aspects of the <i>-ome</i> at hand.
</li>
</ol>
<p>
While grand in scope, these approaches carry difficulties that to me
appear unique. As stated above, <i>-omic</i> information is acquired
through high-throughput techniques, which means that they generate
very large amounts of data. Of major concern is actually linking
this data to biological function. In other words, scientists must
answer whether the correlations in the data actually allow us to
predict the behavior of a cell or tissue. As might be expected,
generating a complex data set such as is found in <i>-omics</i> studies
can be quite impressive at first glance. But this complexity may
hide the fact that no biologically relevant conclusions can be drawn
from it.
</p>

<p>
The authors specifically enumerate four limitations that could
adversely affect <i>-omics</i> studies:
</p>

<ol class="org-ol">
<li>Technical limitations
</li>
<li>Limitations in the experimental design (such as heavy reliance on
assumptions)
</li>
<li>Conceptual limitations
</li>
<li>Limitations in the analysis
</li>
</ol>
<p>
Particularly, the conceptual limitations struck me as
intriguing. The authors offered the example of genomics studies in
which the notion of a "gene" is not currently a well-defined
concept. When the concepts underpinning an entire collection of
measurements is unclear, we should question whether the measured
data has the meaning that we think it does.
</p>

<p>
Overall, I found that this commentary provided an interesting and
sobering view of one particular approach to studying biology. It's
interesting because I suspect that many important biological
questions in the near future will come from taking an integrated and
systems perspective. This perspective will require high-throughput
techniques that carry the same limitations that <i>-omics</i> studies
have.
</p>
    </div>
    </article><article class="h-entry post-text"><header><h1 class="p-name entry-title"><a href="posts/wikibooks.html" class="u-url">Customized Wikibooks are awesome</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">
                Kyle M. Douglass
            </span></p>
            <p class="dateline"><a href="posts/wikibooks.html" rel="bookmark"><time class="published dt-published" datetime="2014-02-15T00:00:00+01:00" title="2014-02-15 00:00">2014-02-15 00:00</time></a></p>
                <p class="commentline">
        
    <a href="posts/wikibooks.html#disqus_thread" data-disqus-identifier="cache/posts/wikibooks.html">Comments</a>


        </p>
</div>
    </header><div class="p-summary entry-summary">
    <p>
Last night the internet failed in our apartment and is still
down. Because of this, I ran into the office today (a Saturday) to
download a bunch of files to do some much-needed background reading
on some topics.
</p>

<p>
This lead me to an interesting question, one which I had not yet
thought of: how can I download Wikipedia pages in a format that
makes for easy offline access?
</p>

<p>
A very quick Google search gave me a surprising answer: make your
own book from Wikipedia pages!
</p>

<p>
Here's how it works. (I'll skip the actual details and <a href="https://en.wikipedia.org/wiki/Help:Books">let you read
it straight from Wikipedia's help files</a>. However, I will explain the
process of making your own book.) After you activate the book
creater, you simply navigate to any Wikipedia page and click the
button on the top of the page that asks if you want to add the
current page to your book. Do this for as many pages as you'd like.
</p>

<p>
When you're ready to access your book, you can do it online, or
export it to a number of formats for offline viewing, such as PDF or
an Open Document format. I thought that the PDF was beautufilly
rendered. It even included all reference information.
</p>

<p>
There are other options with this tool too, such as sorting your
articles into chapters or naming your book. I believe that you can
share it with others as well.
</p>

<p>
Cheers to you, Wikipedia, for creating an awesome and useful
feature!
</p>
    </div>
    </article><article class="h-entry post-text"><header><h1 class="p-name entry-title"><a href="posts/alignment.html" class="u-url">The art of alignment</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">
                Kyle M. Douglass
            </span></p>
            <p class="dateline"><a href="posts/alignment.html" rel="bookmark"><time class="published dt-published" datetime="2014-01-27T00:00:00+01:00" title="2014-01-27 00:00">2014-01-27 00:00</time></a></p>
                <p class="commentline">
        
    <a href="posts/alignment.html#disqus_thread" data-disqus-identifier="cache/posts/alignment.html">Comments</a>


        </p>
</div>
    </header><div class="p-summary entry-summary">
    <p>
I've been working on realigning one of our STORM microscopy setups
and moving the other over the past couple weeks. As a result, I've
been doing a lot more alignment lately than I normally do, which has
got me thinking about alignment in general.
</p>

<p>
I've worked in optics labs for nearly ten years now, but I was never
systematically taught how to do alignments. Eventually, I just
figured it out from asking and watching lots of others do it. This
part of experimental optics seems to be something that's passed down
and shared across generations of PhD students and post-docs, like
some sort of cultural heritage preserved by spoken word.
</p>

<p>
Unfortunately, this fact means that written tutorials or alternative
resources on alignment are scarce. This in turn leads to frustration
for new students who can't easily find someone to show them how to
align a system. I was lucky in my training because both my
undergraduate and graduate programs were in centers dedicated to the
study of optics and optical engineering. Resources were
plentiful. But for a biologist working with a custom microscope
using multiple laser lines, finding someone who is competent in
optics to help realign their system can be highly unlikely.
</p>

<p>
So how can people who don't have a background in optics be trained
in alignment? I'm certain that written tutorials are <i>not</i> the best
tool, since aligning an optics setup is a very hands-on job. But, a
tutorial that makes explicit the principles of alignment might make
a good starting point for others.
</p>

<p>
What are these principles? Below is my rough, initial list that
applies to aligning laser beams. Aligning incoherent systems is
another beast altogether.
</p>

<ol class="org-ol">
<li>The angle of the beam propagation is just as important as the
position of the beam in a plane perpendicular to the table. In
other words, just because a beam is going through a stopped-down
iris doesn't mean it's traveling where you think it is.
</li>
<li>You need to measure the beam position in two planes to determine
its angle of propagation.
</li>
<li>Use "layers of abstraction" to divide sections of the setup into
independent units.
</li>
<li>Use paired mirrors to get enough degrees of freedom to establish
these abstraction layers.
</li>
<li>Design feedback mechanisms into the system. In other words, keep
some irises or alignment marks on the walls in place to aid in
future realignments.
</li>
</ol>
</div>
    </article><article class="h-entry post-text"><header><h1 class="p-name entry-title"><a href="posts/hello-world.html" class="u-url">Hello, world!</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn">
                Kyle M. Douglass
            </span></p>
            <p class="dateline"><a href="posts/hello-world.html" rel="bookmark"><time class="published dt-published" datetime="2013-12-21T00:00:00+01:00" title="2013-12-21 00:00">2013-12-21 00:00</time></a></p>
                <p class="commentline">
        
    <a href="posts/hello-world.html#disqus_thread" data-disqus-identifier="cache/posts/hello-world.html">Comments</a>


        </p>
</div>
    </header><div class="p-summary entry-summary">
    <p>
Well, I've finally got this site off the ground… more or less.
</p>

<p>
This website is the continuation of my blog, <a href="http://quadrule.blogspot.com">My Quad-Ruled Life</a>. On
that site, I explored a number of topics as a graduate student in
optics, while improving my writing as well. As the site progressed
and I matured academically, I found that I was blogging less often,
but still used the blog as a means of exploring ideas. Additionally,
I desired a place to present notes and code to the public in an
easily accessible and open manner.
</p>

<p>
So, after discovering <a href="http://getnikola.com/">Nikola</a> and <a href="http://pages.github.com/">GitHub Pages</a>, I quickly hashed
together a fully homemade site that allowed me to keep writing blog
posts and to present some of the fruits of my research freely on the
internet.
</p>

<p>
I hope you'll find this site useful as a learning resource and as a
means to get to know me better as a scientist.
</p>

<p>
-Kyle
</p>
    </div>
    </article>
</div>

        <nav class="postindexpager"><ul class="pager">
<li class="previous">
                <a href="index.html" rel="prev">Newer posts</a>
            </li>
        </ul></nav><script>var disqus_shortname="kmdouglass";(function(){var a=document.createElement("script");a.async=true;a.src="//"+disqus_shortname+".disqus.com/count.js";(document.getElementsByTagName("head")[0]||document.getElementsByTagName("body")[0]).appendChild(a)}());</script>
</div>
        <!--End of body content-->

        <footer id="footer">
            Contents © 2015         <a href="mailto:kyle.douglass@epfl.ch">Kyle M. Douglass</a> - Powered by         <a href="http://getnikola.com" rel="nofollow">Nikola</a>         
            
        </footer>
</div>
</div>


            <script src="assets/js/all-nocdn.js"></script><script>$('a.image-reference:not(.islink) img:not(.islink)').parent().colorbox({rel:"gal",maxWidth:"100%",maxHeight:"100%",scalePhotos:true});</script><!-- fancy dates --><script>
    moment.locale("en");
    fancydates(0, "YYYY-MM-DD HH:mm");
    </script><!-- end fancy dates -->
</body>
</html>
