<p>
As a physcist, I'm a relative newcomer to the biological
sciences. As a result, many approaches to doing science within these
fields are new and really interesting to me.
</p>

<p>
One such idea is the <i>-omics</i> approach to science. Sonja Prohaska
and Peter Stadler provide an insightful and sometimes amusing
intrepetation of <i>-omics</i> studies in an article in the book
<i>Bioinformatics for Omics Data.</i> (The PubMed link to the article is
<a href="http://www.ncbi.nlm.nih.gov/pubmed/21370084">here</a>, but you can find a pdf by Googling "The use and abuse of
-omes.") According to them, <i>-omics</i> refers to the research field of
digesting the pile of data coming from measurements of a particular
<i>-ome</i>, such as the "genome" or "transcriptome." The goal is to
relate the collection of parts within the <i>-ome</i> to biological
function, or at least to determine function by comparing <i>-omes</i> of
two different organisms.
</p>

<p>
The authors explain that <i>-omics</i> approaches have three components,
which I quote from their article:
</p>

<ol class="org-ol">
<li>A suite of (typically high-throughput) technologies address a
well-defined collection of biological objects, the pertinent
<i>-ome</i>.
</li>
<li>Both technologically and conceptually there is a cataloging
effort to enumerate and characterize the individual
objects&#x2013;hopefully coupled with an initiative to make this
catalog available as a database.
</li>
<li>Beyond the enumeration of objects, one strives to cover the
quantitative aspects of the <i>-ome</i> at hand.
</li>
</ol>

<p>
While grand in scope, these approaches carry difficulties that to me
appear unique. As stated above, <i>-omic</i> information is acquired
through high-throughput techniques, which means that they generate
very large amounts of data. Of major concern is actually linking
this data to biological function. In other words, scientists must
answer whether the correlations in the data actually allow us to
predict the behavior of a cell or tissue. As might be expected,
generating a complex data set such as is found in <i>-omics</i> studies
can be quite impressive at first glance. But this complexity may
hide the fact that no biologically relevant conclusions can be drawn
from it.
</p>

<p>
The authors specifically enumerate four limitations that could
adversely affect <i>-omics</i> studies:
</p>

<ol class="org-ol">
<li>Technical limitations
</li>
<li>Limitations in the experimental design (such as heavy reliance on
assumptions)
</li>
<li>Conceptual limitations
</li>
<li>Limitations in the analysis
</li>
</ol>

<p>
Particularly, the conceptual limitations struck me as
intriguing. The authors offered the example of genomics studies in
which the notion of a "gene" is not currently a well-defined
concept. When the concepts underpinning an entire collection of
measurements is unclear, we should question whether the measured
data has the meaning that we think it does.
</p>

<p>
Overall, I found that this commentary provided an interesting and
sobering view of one particular approach to studying biology. It's
interesting because I suspect that many important biological
questions in the near future will come from taking an integrated and
systems perspective. This perspective will require high-throughput
techniques that carry the same limitations that <i>-omics</i> studies
have.
</p>
